### ISSUE TITLE
dotting the Is and crossing the Ts ;)

### ISSUE THREAD

**Issue by RJVB** [zaps166/QMPlay2#713]
A tiny change to make things pedantically correct for building on Mac:

```
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 18342ae5..b3bd9a31 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -1,5 +1,9 @@
 cmake_minimum_required(VERSION 3.16)
-project(QMPlay2)
+if(APPLE)
+    project(QMPlay2 LANGUAGES C CXX OBJCXX)
+else()
+    project(QMPlay2)
+endif()
 
 if (NOT ${CMAKE_GENERATOR} MATCHES "Ninja")
     message(FATAL_ERROR "Unsupported generator ${CMAKE_GENERATOR}. Ninja is the only supported generator.")
```

Usually this is not required because of the standard compiler (clang) will deduce the language from file extensions. That's probably true for GCC too, except that `g++` will fail to build Apple-specific ObjC++ code (even when using libc++ for the C++ runtime).

Using GCC on Mac may seem like a corner case (with good justifications though) but I'm submitting this change anyway because it's officially "the right thing to do". ;)

**Comment by zaps166** [comment]
Why just not `project(QMPlay2 LANGUAGES CXX OBJCXX)` for all OSes? I don't have any `C` files and `OBJCXX` doesn't seem to make problem on Linux :smile:

**Comment by RJVB** [comment]
On Wednesday April 24 2024 08:14:25 Błażej Szczygieł wrote:
>Why just not `project(QMPlay2 LANGUAGES CXX OBJCXX)` for all OSes? I don't have any `C` files and `OBJCXX` doesn't seem to make problem on Linux :smile: 

You're right about the C files of course, but do you have ObjC++ files on Linux?  But why not - I don't think claiming otherwise will have negative side-effects (and if ever it does you can always split things up again).

**Comment by RJVB** [comment]
On a sidewise related note: you're probably aware that I've been building QMPlay2 against Qt 5.9 on my Mac, and now against 5.12 on my Linux rig (where I really could upgrade to 5.15 but have been too lazy to refactor my personal patches).

I understand you prefer to keep the code clean and that 5.15 is probably closest to Qt 6, but I can provide you with the patchfiles (or point you to them) if you think others may be interested.

**Comment by zaps166** [comment]
Qt 5.9 is very old. It's better to build with 5.15.2 and deploy on older machines. I can run Qt 5.15.2 on openSUSE 13.1 (~10.5 years old system) thanks to one simple Qt patch. So it'll run on your Linux for sure.

Again, which macOS version do you have there?

**Comment by RJVB** [comment]
So what's with Qt 5.15.10 on Linux, isn't that the last version in the 5.15 series? I know even my older Linux rig (Ubuntu 14.04 based) can run 15.5.2 from the official repos.

>Again, which macOS version do you have there?

10.9, which officially is limited to Qt 5.8 but I managed to get Qt 5.9 to work by reintroducing a bunch of code upstream stripped. It *might* be possible to get the X11 version of later Qt5 releases to work but that's quite a bit more work than the occasional backport in dependent software. Plus it is of course not ideal for other things.
As you probably remember I *could* update my Mac OS (to 10.13) but have my reasons to stick with this old version, "for now".

AFAIK the Mac is the only platform where Qt have been quite aggressive in dropping support for older releases and that's a shame because Macs are such long-lived computers. 10.13 is theoretically capable to run Qt 5.15 from what I hear, but 10.12 is already limited to Qt 5.11 .

**Comment by zaps166** [comment]
> I don't think claiming otherwise will have negative side-effects (and if ever it does you can always split things up again).

Didn't build on Windows :rofl:

**Comment by zaps166** [comment]
> So what's with Qt 5.15.10 on Linux, isn't that the last version in the 5.15 series? I know even my older Linux rig (Ubuntu 14.04 based) can run 15.5.2 from the official repos.

In my AppImage I still use 5.15.2 (too lazy to update - it'll not change anything). But the newest is 5.15.13 and will be more updates (it's LTS).

---

For now you can revert a few commits to build QMPlay2 on Qt 5.9.

Do you really run QMPlay2 on macOS using X11?

**Comment by RJVB** [comment]
On Thursday April 25 2024 10:49:54 Błażej Szczygieł wrote:
>Didn't build on Windows :rofl: 

Huh, with what error??

> For now you can revert a few commits to build QMPlay2 on Qt 5.9.

Something like that. I revert the commit that bumps the mindep to 5.15, and have separate patch handles the rest. That one probably evolved over time to be different from simply reverting a few commits.

> Do you really run QMPlay2 on macOS using X11?

Nope, for now there's no advantage to it except if for some reason I want to view some video content on my Mac remotely. But I can :)

**Comment by zaps166** [comment]
> Huh, with what error??

IIRC CMake complained about OBJ-C++ on Windows, compilation not started.

> Something like that. I revert the commit that bumps the mindep to 5.15

Btw. Do you have C++17 there?

> Nope, for now there's no advantage to it except if for some reason I want to view some video content on my Mac remotely.

:smile:

**Comment by RJVB** [comment]
>IIRC CMake complained about OBJ-C++ on Windows, compilation not started.

Strange, I hope it did verify whether or not ObjC++ was supported. If it did it would make sense though still be a bit overly protective. And of course it should actually have verified whether or not any ObjC++ was actually being used, too.
I've been reporting some CMake issues to them lately, and can do that with this one too. If you're willing to provide the output so I can make an initial report.

>Btw. Do you have C++17 there?

I do have a newer libc++ . I remember that I did ask you to avoid C++17 a while back because it gave me problems then, but that appears to be the case no longer. Probably because I build the code via a MacPorts port I wrote, basically a meta-buildsystem with hooks to add all the supporting stuff needed to run modern apps on older OSes.
